### PR TITLE
Support charms using charmcraft extensions

### PR DESCRIPTION
**Pull request by weiiwang01**
Charmcraft now supports extensions (https://discourse.charmhub.io/t/manage-extensions/13788), similar to Snapcraft extensions. These are built-in Python modules that can modify the contents of `charmcraft.yaml`, reducing boilerplate for charms that share common functionalities. Consequently, extensions can alter certain charm metadata, such as resources.

In the pull request, update the charming actions to apply extensions when necessary before getting charm metadata.

### CODE REVIEW

**Comment by cbartz** [comment]
@addyess Hi, could someone from the team please review this PR? Thanks.

**Comment by addyess** [comment]
@cbartz i don't have much of an opinion on this PR.  I'm not sure i'm the right reviewer (nor do i have any power here)

**Comment by weiiwang01** [comment]
> @cbartz i don't have much of an opinion on this PR. I'm not sure i'm the right reviewer (nor do i have any power here)

Sorry, my memory must have failed me, I thought you were the owner of this repository 🤦

**Comment by cbartz** [comment]
@simskij You seem to be the owner of the repo, could you please review the PR? It is currently not possible to use the `upload-charm` action correctly with the `flask-framework` extension (see https://github.com/canonical/github-runner-webhook-router/actions/runs/9363146330/job/25773294254#step:6:20, no oci resource is set, compared to https://github.com/canonical/indico-operator/actions/runs/9480783604/job/26122208282#step:6:477, where the oci resource is set).

File: src/services/charmcraft/charmcraft.ts [hunk 1]
```diff
@@ -164,16 +164,29 @@ class Charmcraft {
     };
   }
 
-  _read() {
+  _readMetadata(): Metadata {
```
Comment (simskij) [inline src/services/charmcraft/charmcraft.ts:167]: this is typescript, we should use `private readMetadata` rather than `_readMetadata`

**Review by simskij — CHANGES_REQUESTED** [review]

File: src/services/charmcraft/charmcraft.ts (hunk 1 above)
Comment (weiiwang01) [reply src/services/charmcraft/charmcraft.ts]: Gotcha, updated, thanks!

**Review by simskij — APPROVED** [review]

**Review by simskij — APPROVED** [review]

**Pull request merged**